dipy/denoise/patch2self.py

```diff
@@ -0,0 +1,273 @@
+        held-out volume.
+
+    y : ndarray
+        Array of patches corresponding to the volume that is used a target for
+        denoising.
+    """
+    # Delete the target volume
+    mask = np.zeros(train.shape[0])
+    mask[vol_idx] = 1
+    cur_x = train[mask == 0]
+    cur_x = cur_x.reshape(((train.shape[0]-1)*train.shape[1],
+                           train.shape[2]))
+
+    # Center voxel of the selected block
+    y = train[vol_idx, train.shape[1]//2, :]
+    return cur_x, y
+
+
+def _vol_denoise(train, vol_idx, model, data, alpha):
+    """ Denoise a single 3D volume using a train and test phase.
+
+    Parameters
+    ----------
+    train : ndarray
+        Array of all 3D patches flattened out to be 2D.
+
+    vol_idx: int
+        The volume number that needs to be held out for training.
+
+    model: str
+        Corresponds to the object of the regressor being used for
+        performing the denoising.
+
+    data: ndarray
+        The 4D noisy DWI data to be denoised.
+
+    alpha: float, optional
+        Regularization parameter only for ridge regression model.
+        default: 1.0
+
+    Returns
+    --------
+    model prediction : ndarray
+        Denoised array of all 3D patches flattened out to be 2D corresponding
+        to the held out volume `vol_idx`.
+
+    """
+    # To add a new model, use the following API
+    # We adhere to the following options as they are used for comparisons
+    if model.lower() == 'ols':
+        model = linear_model.LinearRegression(copy_X=False)
+
+    elif model.lower() == 'ridge':
+        model = linear_model.Ridge(copy_X=False, alpha=alpha)
+
+    elif model.lower() == 'lasso':
+        model = linear_model.Lasso(copy_X=False, max_iter=50)
+
+    else:
+        raise ValueError('Model not supported. ',
+                         'Choose from: ols, ridge or lasso',
+                         data.shape)
+
+    cur_x, y = _vol_split(train, vol_idx)
+    model.fit(cur_x.T, y.T)
+
+    return model.predict(cur_x.T).reshape(data.shape[0], data.shape[1],
+                                          data.shape[2])
+
+
+def _extract_3d_patches(arr, patch_radius):
+    """ Extract 3D patches from 4D DWI data.
+
+    Parameters
+    ----------
+    arr : ndarray
+        The 4D noisy DWI data to be denoised.
+
+    patch_radius : int or 1D array
+        The radius of the local patch to be taken around each voxel (in
+        voxels).
+
+    Returns
+    --------
+    all_patches : ndarray
+        All 3D patches flattened out to be 2D corresponding to the each 3D
+        volume of the 4D DWI data.
+
+    """
+    if isinstance(patch_radius, int):
+        patch_radius = np.ones(3, dtype=int) * patch_radius
+    if len(patch_radius) != 3:
+        raise ValueError("patch_radius should have length 3")
+    else:
+        patch_radius = np.asarray(patch_radius, dtype=(int))
+    patch_size = 2 * patch_radius + 1
+
+    dim = arr.shape[-1]
+
+    all_patches = []
```

Comment:
I'm pretty sure you could preallocate this, making it go faster and (more importantly) be less memory-intensive.

```suggestion
    all_patches = np.zeros((np.prod(arr.shape[:3]), 1, dim))
```

Reply:
Then you would have to keep count and allocate this row-by-row in the loop below.

Reply:
Yes... I tried doing this.. and also profiled the memory with `mem-prof`. Dint really change much...

```
    all_patches = np.zeros((arr.shape[0]*arr.shape[1]*arr.shape[2],
                            np.prod(patch_size), dim))

    # loop around and find the 3D patch for each direction
    for i in range(patch_radius[0], arr.shape[0] -
                       patch_radius[0], 1):
            for j in range(patch_radius[1], arr.shape[1] -
                           patch_radius[1], 1):
                for k in range(patch_radius[2], arr.shape[2] -
                               patch_radius[2], 1):
                    # Shorthand for indexing variables:
                    ix1 = i - patch_radius[0]
                    ix2 = i + patch_radius[0] + 1
                    jx1 = j - patch_radius[1]
                    jx2 = j + patch_radius[1] + 1
                    kx1 = k - patch_radius[2]
                    kx2 = k + patch_radius[2] + 1

                    X = arr[ix1:ix2,
                            jx1:jx2,
                            kx1:kx2].reshape(np.prod(patch_size), dim)

                    b = arr.shape[1]
                    c = arr.shape[2]

                    index0 = (i*b*c) + (j*c) + (k)
                    all_patches[index0, :, :] = X

    return np.transpose(all_patches)
```